cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2426,25 +2449,33 @@ std::shared_ptr<VectorFunction> MakeIndicesNonZeroFunction(std::string name,
                                                            const FunctionDoc* doc) {
   auto func = std::make_shared<VectorFunction>(name, Arity::Unary(), doc);
 
-  for (const auto& ty : NumericTypes()) {
-    VectorKernel kernel;
-    kernel.exec = IndicesNonZeroExec;
-    kernel.null_handling = NullHandling::OUTPUT_NOT_NULL;
-    kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
-    kernel.output_chunked = false;
-    kernel.can_execute_chunkwise = false;
-    kernel.signature = KernelSignature::Make({InputType(ty->id())}, uint64());
-    DCHECK_OK(func->AddKernel(kernel));
-  }
-
-  VectorKernel boolkernel;
-  boolkernel.exec = IndicesNonZeroExec;
-  boolkernel.null_handling = NullHandling::OUTPUT_NOT_NULL;
-  boolkernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
-  boolkernel.output_chunked = false;
-  boolkernel.can_execute_chunkwise = false;
-  boolkernel.signature = KernelSignature::Make({boolean()}, uint64());
-  DCHECK_OK(func->AddKernel(boolkernel));
+  VectorKernel kernel;
+  kernel.null_handling = NullHandling::OUTPUT_NOT_NULL;
+  kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
+  kernel.output_chunked = false;
+  kernel.can_execute_chunkwise = false;
+
+  auto AddKernels = [&](const std::vector<std::shared_ptr<DataType>>& types) {
+    for (const std::shared_ptr<DataType>& ty : types) {
+      kernel.signature = KernelSignature::Make({InputType::Array(ty)}, uint64());
+      kernel.exec = IndicesNonZeroExec;
```

Comment:
This line can probably be moved out of this lambda?

Reply:
done